libgo/procmgr/manager.go

```diff
@@ -0,0 +1,370 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/gob"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+)
+
+type callback struct {
+	function, data uintptr
+}
+
+type manager struct {
+	ipcConns struct {
+		parent, child net.Conn
+	}
+	writeQueue struct {
+		items []interface{}
+		lock  sync.Mutex
+		feed  chan struct{}
+	}
+	ourEnv struct {
+		vars []string
+		once sync.Once
+	}
+	pendingCallbacks struct {
+		items map[uint64]callback
+		lock  sync.Mutex
+	}
+	nextIPid uint64
+}
+
+var _ io.Closer = (*manager)(nil)
+
+type spawnRequest struct {
+	iPid     uint64
+	args     []string
+	extraEnv []string
+	timeout  time.Duration
+}
+
+var _ = func() struct{} {
+	gob.Register((*spawnRequest)(nil))
+	return struct{}{}
+}()
+
+type exitStatus struct {
+	iPid      uint64
+	pid       int
+	exitCode  int
+	output    []byte
+	execStart time.Time
+	execEnd   time.Time
+}
+
+var _ = func() struct{} {
+	gob.Register((*exitStatus)(nil))
+	return struct{}{}
+}()
+
+var aliveManagers = struct {
+	items map[*manager]struct{}
+	lock  sync.Mutex
+}{}
+
+func (m *manager) Close() (err error) {
+	for _, conn := range [2]net.Conn{m.ipcConns.parent, m.ipcConns.child} {
+		if conn != nil {
+			if errCl := conn.Close(); errCl != nil && err == nil {
+				err = errCl
+			}
+		}
+	}
+
+	aliveManagers.lock.Lock()
+	defer aliveManagers.lock.Unlock()
+
+	delete(aliveManagers.items, m)
+
+	return
+}
+
+func (m *manager) beforeFork() error {
+	fds, errSP := socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM, 0)
+	if errSP != nil {
+		return errSP
+	}
+
+	defer closeFdIfSet(&fds[0])
+	defer closeFdIfSet(&fds[1])
+
+	syscall.SetNonblock(fds[0], true)
+	syscall.SetNonblock(fds[1], true)
+
+	f0 := os.NewFile(uintptr(fds[0]), "")
+	defer f0.Close()
+	fds[0] = -1
+
+	f1 := os.NewFile(uintptr(fds[1]), "")
+	defer f1.Close()
+	fds[1] = -1
+
+	c0, errFC0 := net.FileConn(f0)
+	if errFC0 != nil {
+		return errFC0
+	}
+
+	defer closeConnIfSet(&c0)
+
+	c1, errFC1 := net.FileConn(f1)
+	if errFC1 != nil {
+		return errFC1
+	}
+
+	defer closeConnIfSet(&c1)
+
+	m.ipcConns.parent = c0
+	c0 = nil
+
+	m.ipcConns.child = c1
+	c1 = nil
+
+	return nil
+}
+
+func (m *manager) afterFork(child bool) {
+	if child {
+		m.ipcConns.parent.Close()
+	} else {
+		m.ipcConns.child.Close()
+	}
+}
+
+func (m *manager) messagesLoop() {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	readLoopDone := make(chan struct{})
+
+	go m.readLoop(&wg, readLoopDone)
```

Comment:
@nilmerg Any idea why this line fires, but the called function doesn't?

Reply:
@lippserd Btw that's a good example for the fact that it's much harder to make just one part web-scale instead of the whole thing "at once".

Reply:
Solved this problem by spawning two threads.